<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8"><meta name="viewport" content="width=device-width,initial-scale=1.0,minimum-scale=1.0"><meta name="google-site-verification" content="4cR7WuQWEtkbu8ThfxGkT1vACiVi5Q_8ykfkRudOnTU" />

<link rel="shortcut icon" href="/images/logo.png">
<link rel="stylesheet" type="text/css" href="/css/common.css"><script type="text/javascript" src="/js/jquery-1.9.0.min.js"></script>
<script type="text/javascript" src="/js/google-code-prettify/prettify.js"></script>
<link  type="text/css" rel="stylesheet" href="/js/google-code-prettify/prettify.css"/>
<script>
  $(function(){
    // google code prettifyの有効化
    $("pre").addClass("prettyprint");
    function init(event){
      prettyPrint();
    }
    if(window.addEventListener)window.addEventListener("load",init,false);
    else if(window.attachEvent)window.attachEvent("onload",init);
    
    $(".to-top").click(function() {
      // ページの一番上までスクロールさせます。
      $('body, html').animate({scrollTop: 0}, 300, 'linear');;
    });
  });
</script>

<script type="text/javascript" language="javascript">    var vc_pid = "886663612";
</script><script type="text/javascript" src="//aml.valuecommerce.com/vcdal.js" async></script>

<title>Let's understand how Perl exception handling works - Perl Zemi|Perl Installation, Perl Tutorial, Many Examples</title>
<meta name="description" content="Perl allows you to use  exception handling  for error notification. Exception handling is very easy. All you have to do is learn how to raise an exception and how to catch it.">
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:site" content="@perlzemi" />
<meta name="twitter:title" content="Let's understand how Perl exception handling works - Perl Zemi|Perl Installation, Perl Tutorial, Many Examples" />
<meta name="twitter:description" content="Perl allows you to use  exception handling  for error notification. Exception handling is very easy. All you have to do is learn how to raise an exception and how to catch it." />
<meta name="twitter:image" content="https://en.perlzemi.com//images/twitter_card_large_kaeru.png" />


  </head>
  <body>
    <div class="container">
      <div class="header">
        <div class="header_main">  <h1>    <a href="/"><img src="/images/logo.png">Perl Zemi</a>
</h1>
</div>

      </div>
      <div class="main">
        <div class="content">
          <div class="entry">
  <div class="top">
    
  </div>
  <div class="middle">
    <ol class = "bread" itemscope itemtype = "http://schema.org/BreadcrumbList">  <li itemprop = "itemListElement" itemscope itemtype = "http://schema.org/ListItem">    <a itemprop="item" href="/">        <span itemprop = "name">Perl </span></a>
    <meta itemprop = "position" content = "1" />
  </li>
  &#8250;
  <li itemprop = "itemListElement" itemscope itemtype = "http://schema.org/ListItem">      <span itemprop = "name">here </span>
  </li>
</ol>
<h2><a href="/blog/20100411127089.html">Let's understand how Perl exception handling works</a></h2>
<p>
  Perl allows you to use <strong> exception handling </strong> for error notification. Exception handling is very easy. All you have to do is learn how to raise an exception and how to catch it.
</p>
<div style="width:calc(100% - 30px);margin:10px auto;">
  <script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-4525414114581084"
       crossorigin="anonymous"></script>
  <!-- 最初の段落下 - ディスプレイ 横長 レスポンシブ -->
  <ins class="adsbygoogle"
       style="display:block"
       data-ad-client="ca-pub-4525414114581084"
       data-ad-slot="2828858335"
       data-ad-format="auto"
       data-full-width-responsive="true"></ins>
  <script>
       (adsbygoogle = window.adsbygoogle || []).push({});
  </script>
</div>


<h3>Raise an exception</h3>
<p>
  Use the die function to raise an exception. When die is executed, the program will exit with an error status. You can pass a string to die.
</p>
<pre>
# Raise an exception
die $message;
</pre>
<p>
  Let's write an example that raises an exception. I am trying to raise an exception if the value is not a positive integer in the function. The program will display a message and exit.
</p>
<pre>
# Check value. Since it is not a positive integer, an exception will be thrown and the program will exit.
is_int('a');

sub is_int {
  my $num = shift;
  
  # Throw an exception if the value is not a positive integer
  die "\" $num\"must be number" unless $num =~ /^\d+$/;
}
</pre>
<h3>Catch exceptions</h3>
<p>
  You can catch an exception if you don't want the program to terminate when the called function raises an exception. Use the <a href="/blog/20100312126967.html">eval</a> block to catch the exception.
</p>
<pre>
eval {Handling that raises an exception};
</pre>
<p>
  When an exception occurs, the message specified by the die function is stored in a predefined variable called $@. Write the processing when an exception occurs using this $@.
</p>
<pre>
if ($@) {
    # What to do if an exception occurs
}
</pre>
<p>
  If an exception occurs, it means that an error has occurred, so in most cases it will not work if you continue processing. The motivation to catch the exception is when there is a possibility of recovering from the error by retrying (in the case of timeout etc.), or when the program ends as it is, garbage such as temporary files will remain, so post-processing is necessary. For example, the case.
</p>
<p>
  This is an example to catch the exception that occurred in the is_int function.
</p>
<pre>
# Catch exceptions
eval {is_int('a')};

# What to do if an exception occurs
if ($@) {
  print "Exception occur: $@";
}

sub is_int {
  my $num = shift;
  
  # Throw an exception if the value is not a positive integer
  die "\" $num\"must be number." unless $num =~ /^\d+$/;
}
</pre>
<h3>When raising an exception in a module</h3>
<p>
  If you want to create a module and raise an exception in the function in it, use the croak function of the Carp module. croak calls die internally. The difference from die is that croak includes the calling line number in the exception message. This allows the user of the module to know where the function or method was called.
</p>
<pre>
package Your Module;

use Carp 'croak';

sub func {
  
  # Raise an exception
  croak "Error message";
}
</pre>

  </div>
  <div class="bottom">
    <h3>Related Informatrion</h3>

<div style="margin:10px 10px;"><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-4525414114581084"
     crossorigin="anonymous"></script>
<ins class="adsbygoogle"
     style="display:block"
     data-ad-format="autorelaxed"
     data-ad-client="ca-pub-4525414114581084"
     data-ad-slot="9163995495"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
</div>

  </div>
</div>

        </div>
        <div class="side">
          <div class="side_list">  <div class="side_list_body">    <ul>
      <li>
        <div class="side_list_image">        </div>
      </li>
    </ul>
  </div>
</div>

        </div>
      </div>
      <div class="footer">
        <div class="perlri_link">  <a rel="nofollow" href="javascript:void(0)">Perl Club</a>
</div>

      </div>
    </div>
  </body>
</html>
